packages/devtools_flutter/lib/src/config.dart

```diff
@@ -10,28 +11,111 @@ import 'screen.dart';
+    return MaterialPageRoute(
+      settings: settings,
+      builder: (_) {
+        return DevToolsScaffold.withChild(
+          child: const Center(
+            child: Text('Sorry, this page was not found'),
+          ),
+        );
+      },
+    );
   }
 
-  // The mapping from routes to the collection of screens to show in the app.
-  //
-  // The /connect route will be a dependency for all the other routes.
-  final Map<String, List<Screen>> _routeToTabs = const {
-    '/': [ConnectScreen()],
-    '/connected': [
-      EmptyScreen.inspector,
-      EmptyScreen.timeline,
-      EmptyScreen.performance,
-      EmptyScreen.memory,
-      EmptyScreen.logging,
-    ],
+  /// The routes that the app exposes.
+  final Map<String, WidgetBuilder> routes = {
+    '/': (_) => Initializer(
+          builder: (_) => const DevToolsScaffold(
+            tabs: [
+              EmptyScreen.inspector,
+              EmptyScreen.timeline,
+              EmptyScreen.performance,
+              EmptyScreen.memory,
+              EmptyScreen.logging,
+            ],
+          ),
+        ),
+    '/connect': (_) => DevToolsScaffold.withChild(child: ConnectScreenBody()),
   };
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      theme: ThemeData.light(),
+      onGenerateRoute: generateRoute,
+    );
+  }
+}
+
+/// Widget that requires business logic to be loaded before building its
+/// [builder].
+///
+/// See [_InitializerState.build] for the logic that determines whether the
+/// business logic is loaded.
+///
+/// Use this widget to wrap pages that require [service.serviceManager] to be
+/// connected. As we require additional services to be available, add them
+/// here.
+class Initializer extends StatefulWidget {
+  const Initializer({Key key, @required this.builder})
+      : assert(builder != null),
+        super(key: key);
+
+  /// The builder for the widget's children.
+  ///
+  /// Will only be built if [_InitializerState._loaded] is true.
+  final WidgetBuilder builder;
+
+  @override
+  _InitializerState createState() => _InitializerState();
+}
+
+class _InitializerState extends State<Initializer> {
+  @override
+  Widget build(BuildContext context) {
+    final loaded = service.serviceManager.hasConnection;
+
+    // If the loading is not completed, navigate to the connection page to
+    // connect to a VM service.
+    if (!loaded && ModalRoute.of(context).isCurrent) {
+      WidgetsBinding.instance.addPostFrameCallback((_) {
+        Navigator.of(context).pushNamed('/connect').then((_) {
+          // Generally, empty setState calls in Flutter should be avoided.
+          // However, serviceManager is an implicit part of the state
+          // of Initializer. This setState call is alerting the widget of a change
+          // in the serviceManager's state.
+          setState(() {});
+        });
+      });
+    }
+
+    return loaded ? widget.builder(context) : const SizedBox();
```

Comment:
Why are we returning a `SizedBox()` in the not loaded case? Maybe add a comment to explain.

Reply:
SizedBox is the most generic no-op widget it's possible to build.

Reply:
I updated the comments to explain what its purpose is here.

In general, `const SizedBox()` means a widget with no size and no contents.  Before the list comprehension and list conditional features were added in Dart 2.3, SizedBox was the most efficient way of creating a list with members that only showed up occasionally:

```
return Column(children: [
  Text('header'),
  widget.body != null ? widget.body : const SizedBox(),
]);
```

Now it's not needed in those cases, but in situations like this it is still useful.